roomserver/internal/input/input_latest_events.go

```diff
@@ -125,7 +124,8 @@ func (u *latestEventsUpdater) doUpdateLatestEvents() error {
 	// then start with an empty set - none of the forward extremities
 	// that we knew about before matter anymore.
 	oldLatest := []types.StateAtEventAndReference{}
-	if !u.stateAtEvent.Overwrite {
+	if !u.rewritesState {
+		u.oldStateNID = u.updater.CurrentStateSnapshotNID()
```

Comment:
Whilst this makes sense, why did you change it so it's only set if we don't rewrite state? I think `updater.CurrentStateSnapshotNID()` just returns an int - the `New...` function did the DB query so it's not even an optimisation. Maybe to act as a guard against accidentally fiddling with it and calculating deltas against things we shouldn't?

Reply:
The idea is that, if we are rewriting state, we *want* the state snapshot NID to be zero because that means that the deltas produced later on will be against an empty snapshot (effectively a complete rewrite, where `AddsStateEventIDs` will contain the entire new state).

If we aren't rewriting state, we want to know what the current state is so that we produce a real delta.